packages/cosmic-swingset/lib/ag-solo/vats/lib-wallet.js

```diff
@@ -61,120 +61,96 @@ export async function makeWallet(
     const {
-      meta: { purseName0, purseName1, instanceId },
+      meta: { instanceId, ...rest },
       offerRules,
     } = dateToOfferRec.get(date);
 
-    const purse0 = petnameToPurse.get(purseName0);
-    const purse1 = petnameToPurse.get(purseName1);
-
-    const {
-      payoutRules: [
-        {
-          units: { extent: extent0, assayId: assayId0 },
-        },
-        {
-          units: { extent: extent1, assayId: assayId1 },
-        },
-      ],
-    } = offerRules;
-
-    const instanceHandleP = E(registrar).get(instanceId);
-    const regAssay0P = E(registrar).get(assayId0);
-    const regAssay1P = E(registrar).get(assayId1);
-    const purseAssay0P = E(purse0).getAssay();
-    const purseAssay1P = E(purse1).getAssay();
-    const purseUnit0P = E(purseAssay0P).makeUnits(extent0 || 0);
-    const purseUnit1P = E(purseAssay0P).makeUnits(extent1 || 0);
+    // Collapse purseName0, purseName1, ... into purseNames
+    const purseNames = [];
+    for (let i = 0; i < offerRules.payoutRules.length; i += 1) {
+      const purseName = rest[`purseName${i}`];
+      if (purseName === undefined) {
+        break;
+      }
+      purseNames.push(purseName);
+    }
+    const rawPurses = purseNames.map(purseName =>
+      petnameToPurse.get(purseName),
+    );
 
-    const [
-      instanceHandle,
-      regAssay0,
-      regAssay1,
-      purseAssay0,
-      purseAssay1,
-      purseUnit0,
-      purseUnit1,
-    ] = await Promise.all([
-      instanceHandleP,
-      regAssay0P,
-      regAssay1P,
-      purseAssay0P,
-      purseAssay1P,
-      purseUnit0P,
-      purseUnit1P,
-    ]);
+    const { payoutRules } = offerRules;
 
-    // =====================
-    // === AWAITING TURN ===
-    // =====================
+    const instanceP = E(registrar)
+      .get(instanceId)
+      .then(instanceHandle => E(zoe).getInstance(instanceHandle));
+    const regAssaysAP = payoutRules.map(({ units: { assayId } }) =>
+      E(registrar).get(assayId),
+    );
+    const regAssaysP = Promise.all(regAssaysAP);
+    const regUnitsP = Promise.all(
+      payoutRules.map(({ units: { extent } }, i) =>
+        E(regAssaysAP[i]).makeUnits(extent || 0),
+      ),
+    );
+    const purseAssaysP = Promise.all(
+      rawPurses.map(purse => E(purse).getAssay()),
+    );
 
-    const {
-      terms: {
-        assays: [contractAssay0, contractAssay1],
+    const [
+      {
+        terms: { assays: contractAssays },
+        publicAPI,
       },
-      publicAPI,
-    } = await E(zoe).getInstance(instanceHandle);
+      regAssays,
+      regUnits,
+      purseAssays,
+    ] = await Promise.all([instanceP, regAssaysP, regUnitsP, purseAssaysP]);
 
     // =====================
     // === AWAITING TURN ===
     // =====================
 
+    // Order the purses by registered assays.
+    const purses = regAssays.map(regAssay => {
+      const i = purseAssays.indexOf(regAssay);
+      if (i < 0) {
+        return undefined;
```

Comment:
Is it acceptable for some members of regAssays to be missing from purses?

Reply:
Yes, I think it is.  Notably, the autoswap liquidity assay may not necessarily correspond to a purse.

Reply:
But I will fix a minor issue: if there are multiple purses for the same assay, they should not be reused.